### PR TITLE
Add an output directory to create_sub_workflow. This provides a user …

### PR DESCRIPTION
**Pull request by cianciosa**
…specified directory to stage output files to.

### CODE REVIEW

File: framework/src/services.py [hunk 1]
```diff
@@ -2366,12 +2380,12 @@ def create_sub_workflow(self, sub_name, config_file, override= None, input_dir=
         ports = sub_conf_new['PORTS']['NAMES'].split()
         comps = [sub_conf_new['PORTS'][p]['IMPLEMENTATION'] for p in ports]
         for c in comps:
-            if not c:
```
Comment (elwasif) [inline framework/src/services.py]: Why is this deleted? This check is needed since sometimes a listed port may not have an implementation, which is fine as long as the driver doesn't use it

Comment (cianciosa) [reply framework/src/services.py]: Thats a good question. Probably shouldn't be.

**Comment by cianciosa** [comment]
This should be fixed in f9a8014c3be0f155631ff978480d1772d1e15f71

File: framework/src/services.py [hunk 2]
```diff
@@ -1649,6 +1650,19 @@ def stage_output_files(self, timeStamp, file_list, keep_old_files = True):
...
+                                  outprefix, keep_old=keep_old_files)
+            except Exception, e:
+                self._send_monitor_event('IPS_STAGE_OUTPUTS',
+                                         'Files = ' + str(file_list) + \
+                                         ' Exception raised : ' + str(e),
+                                         ok='False')
+                self.exception('Error in stage_output_files()')
+                raise
+
         # Store plasma state files into $SIM_ROOT/history/plasma_state
```
Comment (elwasif) [inline framework/src/services.py]: So this will copy all output files from all components in the sub flow in every step where stage_output_files() is  called to a single OUTPUT_DIR, is this really the right  design?. What if the sub-flow has several components and runs for tens of steps or more - do we want to copy all of these files and make them available to the super-flow?

Comment (cianciosa) [reply framework/src/services.py]: In the usage scenario I envision, the super workflow doesn't know or care what components staged output files. What I really need is just the current plasma state of the sub workflow.

Attached is some slides explaining how the workflows couple.

[Nested workflow use cases.pptx](https://github.com/HPC-SimTools/IPS-framework/files/1910819/Nested.workflow.use.cases.pptx)

Comment (elwasif) [reply framework/src/services.py]: Check this call https://github.com/HPC-SimTools/IPS-framework/blob/d39232eb1e6c44b41f36e6f35ccc2011271e6cdd/framework/src/services.py#L1582 for a way to access subflow output from the super flow.

**Comment by dlg0** [comment]
My understanding of using existing IPS workflows as components within an "super" workflow was that the sub-workflow was treated as any other regular component. However, this pull request seems to indicate that this is not the case, i.e., one cannot stage the state files from the super-workflow to the component (here a sub-workflow), is that correct?

Can we alter this behavior such that the sub-workflows are treated as first-class components? (If not, why not).

**Comment by elwasif** [comment]
@dlg0 The current design assumes that output files from subflows are the files specified in the subflow's driver OUTPUT_FILES config parameter, and they are retrieved from the subflow's driver's work subdirectory via the call to stage_subflow_output_files(). The underlying assumption is that anything else in the subflow is "tmp" files that is of no interest to the super flow. If we have a use case that requires a different behaviour we can alter this call, or add another service. We just need to define what behavior we want to support.

**Comment by cianciosa** [comment]
Won't stage_subflow_output_files() create a conflict when there are multiple parallel instances of the same sub workflow?

**Comment by elwasif** [comment]
yes. So it wasn't really designed with "ensemble of subflows" in mind. ON ecan resolve the conflict by putting each set of output files into its own subdirectory, with a name derived from the subflow name for example.

**Comment by dlg0** [comment]
Is there a `stage_subflow_input_files()` also? i.e., these would be the INPUT_FILES specified in the super workflow config file under the component defining the subworkflow.

And how does `stage_subflow_output_files()` differ from `stage_output_files()`? Is it as simple as overloading `stage_input_files()` and `stage_output_files()` to handle subworkflows AND regular components to make subworkflows first class components?

**Comment by elwasif** [comment]
@dlg There's really no need for stage_subflow_input_files(), since the input files for the sub flow are copied into a subdirectory at subflow creation time, and the sub flow config file is edited to point to that directory. So in the sub flow, calling stage_input() will get the files from this new directory.

The distinction between `stage_subflow_output_files()`  and `stage_output_files()` stems from the need of the sub flow "bridge component" to gather output files from the sub flow before it can do anything with them in the "upper level" flow.

**Comment by cianciosa** [comment]
Closing since this is no longer needed.
